ContestModel/src/org/icpc/tools/contest/model/internal/account/PublicContest.java

```diff
@@ -0,0 +1,185 @@
+ * <li>Clarifications (except broadcasts)</li>
+ * <li>Judgemewnts and commentary after the freeze</li>
+ * <li>Test_data</li>
+ * <li>Runs</li>
+ * <li>Commentary</li>
+ * </ul>
+ */
+public class PublicContest extends Contest {
+	protected String username;
+
+	protected List<IProblem> problems = new ArrayList<IProblem>();
+
+	public PublicContest() {
+		// do nothing
+	}
+
+	@Override
+	public void add(IContestObject obj) {
+		IContestObject.ContestType cType = obj.getType();
+
+		switch (cType) {
+			// all of these are fully public
+			case CONTEST:
+			case JUDGEMENT_TYPE:
+			case MAP_INFO:
+			case START_STATUS:
+			case AWARD:
+				super.add(obj);
+				return;
+			case PROBLEM: {
+				if (getState().getStarted() == null)
+					problems.add((IProblem) obj);
+				else
+					super.add(obj);
+				return;
+			}
+			case STATE: {
+				IState state = (IState) obj;
+				super.add(state);
+				// TODO out of order!
+				if (state.getStarted() != null && getProblems().length == 0) {
+					for (IProblem p : problems) {
+						super.add(p);
+					}
+				}
+				return;
+			}
+			case GROUP: {
+				IGroup group = (IGroup) obj;
+				if (!group.isHidden())
+					super.add(group);
+				return;
+			}
+			case TEAM: {
+				ITeam team = (ITeam) obj;
+				if (!isTeamHidden(team))
+					super.add(team);
+				return;
+			}
+			case ACCOUNT: {
+				// everyone only sees their own account (public has no account)
+				IAccount account = (IAccount) obj;
+				if (username == null || !username.equals(account.getUsername()))
+					return;
+
+				if (account.getPassword() != null) {
+					account = (IAccount) ((Account) account).clone();
+					((Account) account).add("password", null);
+				}
+				super.add(account);
+				return;
+			}
+			case TEAM_MEMBER: {
+				ITeamMember member = (ITeamMember) obj;
+				ITeam team = getTeamById(member.getTeamId());
+				if (!isTeamHidden(team))
+					super.add(member);
+				return;
+			}
+			case SUBMISSION: {
+				ISubmission sub = (ISubmission) obj;
+
+				// hide submissions from outside the contest time
+				int time = sub.getContestTime();
+				if (time < 0 || time >= getDuration())
+					return;
+
+				// hide submissions from hidden teams
+				ITeam team = getTeamById(sub.getTeamId());
+				if (isTeamHidden(team))
+					return;
+
+				super.add(sub);
+				return;
+			}
+			case JUDGEMENT: {
+				IJudgement j = (IJudgement) obj;
+
+				ISubmission s = getSubmissionById(j.getSubmissionId());
+				if (s == null)
+					return;
+
+				// hide judgements for submissions outside the contest time
+				int time = s.getContestTime();
+				if (time < 0 || time >= getDuration())
+					return;
+
+				// or during the freeze
+				if (getFreezeDuration() != null) {
+					int freezeTime = getDuration() - getFreezeDuration();
+					if (time >= freezeTime)
+						return;
+				}
+
+				// don't show judgements for hidden teams
+				if (isJudgementHidden(j))
+					return;
+
+				super.add(j);
```

Comment:
The spec says this is fine, but do we actually want to show the public the exact verdict for submissions before the freeze? Not for this PR, but didn't we talk about it during a CLICS call?

Reply:
Agreed, I thought there was a discussion about a simplified set of judgement types for public. I can't find anything in the spec now though, will add to the agenda.